Wait until cursor is done iterating to mark subscription ready

qualitymanifest opened the pull request:
Fixes https://github.com/JcBernack/meteor-reactive-aggregate/issues/48

- Removed the original `subscription.ready()`
- Changed the `forEach` running on the aggregation cursor to an `each` since this gives you a `null` in place of a document when the cursor is done iterating. When cursor is done iterating, remove errant documents and mark subscription ready
- If aggregation fails, send the client an error so their subscription stops and they can handle that error

Let me know if this is okay, this is my first PR so I hope I'm doing everything right. Thanks

qualitymanifest committed:
Wait until cursor is empty to mark subscription ready

qualitymanifest committed:
Handle errors during aggregation

JcBernack commented:
Nice, thanks for the PR. The `.forEach` was not the cause of the issue though, it is synchronous. As I said in #48 the `_.throttle` is causing the update function to be asynchronous. Anyway, your changes should fix the issue. Have you tested it in your use case?

qualitymanifest commented:
Oh sorry, I should have mentioned that - I tested it and it works as intended. Further, I tried putting an error in my aggregation and the client receives that error in their `onStop` callback.

For what it's worth (in case this gets referenced in the future), I may well be wrong about the `forEach` callback being the problem, but I don't think it was the `_.throttle`:
Before I tried anything else, I tried putting the aggregation function on line 79 in place of the first `update()` so it would run without throttle, and threw in some `console.log`s like
```
Meteor.bindEnvironment(() => {
  // add and update documents on the client
  collection.aggregate(safePipeline).forEach((doc) => {
    console.log("ITERATING CURSOR");
    if (!subscription._ids[doc._id]) {
      subscription.added(clientCollection, doc._id, doc);
    } else {
      subscription.changed(clientCollection, doc._id, doc);
    }
    subscription._ids[doc._id] = subscription._iteration;
  });
  // remove documents not in the result anymore
  _.each(subscription._ids, (iteration, key) => {
    if (iteration != subscription._iteration) {
      delete subscription._ids[key];
      subscription.removed(clientCollection, key);
    }
  });
  subscription._iteration++;
})
console.log("DONE");
```
Same issue, and I would get
```
DONE
ITERATING CURSOR
ITERATING CURSOR
...etc
```

JcBernack commented:
Nice, ok.
